apollo-router/src/cache/storage.rs

```diff
@@ -187,25 +255,24 @@ where
                 .await;
         }
 
-        let mut in_memory = self.inner.lock().await;
-        in_memory.put(key, value);
-        let size = in_memory.len() as u64;
-        tracing::info!(
-            value.apollo_router_cache_size = size,
-            kind = %self.caller,
-            storage = &tracing::field::display(CacheStorageName::Memory),
-        );
+        self.insert_in_memory(key, value).await;
     }
 
-    pub(crate) async fn insert_in_memory(&self, key: K, value: V) {
+    pub(crate) async fn insert_in_memory(&self, key: K, value: V)
+    where
+        V: ValueType,
+    {
+        // Update the cache size and estimated storage size
+        // This is cheaper than trying to estimate the cache storage size by iterating over the cache
+        self.cache_estimated_storage
+            .fetch_add(value.estimated_size().unwrap_or(0) as i64, Ordering::SeqCst);
         let mut in_memory = self.inner.lock().await;
-        in_memory.put(key, value);
-        let size = in_memory.len() as u64;
-        tracing::info!(
-            value.apollo_router_cache_size = size,
-            kind = %self.caller,
-            storage = &tracing::field::display(CacheStorageName::Memory),
-        );
+        if let Some((_, v)) = in_memory.push(key.clone(), value.clone()) {
```

Comment:
the in memory cache lock is held until the end of the function, it should only be held when inserting the old entry.
Maybe fixed with something like:
```rust
let old_entry = {
  self.inner.lock().await.push(key.clone(), value.clone())
};
if let Some((_, v)) = old_entry {
```

Reply:
Fixed in: [e37c87e](https://github.com/apollographql/router/pull/5770/commits/e37c87ebc97cb588dfc1de831760df1d0ce1d68f)